File: kmeans_kd_tree.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <cmath>
#include <limits>
#include <algorithm>
#include <numeric>

struct Point {
    double x, y;

    Point() : x(0), y(0) {}
    Point(double x, double y) : x(x), y(y) {}

    // Sobrecarga del operador de igualdad
    bool operator==(const Point& other) const {
        return x == other.x && y == other.y;
    }
};

struct KDNode {
    Point point;
    int axis;
    KDNode* left;
    KDNode* right;

    KDNode(Point p, int axis) : point(p), axis(axis), left(nullptr), right(nullptr) {}
};

double distance(const Point& p1, const Point& p2) {
    return sqrt(pow(p1.x - p2.x, 2) + pow(p1.y - p2.y, 2));
}

KDNode* buildKDTree(const std::vector<Point>& points, int depth) {
    if (points.empty()) {
        return nullptr;
    }

    int axis = depth % 2; // Alternar entre las dimensiones x e y
    auto cmp = (axis == 0) ? [](const Point& p1, const Point& p2) { return p1.x < p2.x; }
                           : [](const Point& p1, const Point& p2) { return p1.y < p2.y; };

    std::vector<Point> sortedPoints = points;
    std::sort(sortedPoints.begin(), sortedPoints.end(), cmp);
    int medianIndex = sortedPoints.size() / 2;
    KDNode* node = new KDNode(sortedPoints[medianIndex], axis);

    // Construir subárboles recursivamente
    std::vector<Point> leftPoints(sortedPoints.begin(), sortedPoints.begin() + medianIndex);
    std::vector<Point> rightPoints(sortedPoints.begin() + medianIndex + 1, sortedPoints.end());
    node->left = buildKDTree(leftPoints, depth + 1);
    node->right = buildKDTree(rightPoints, depth + 1);

    return node;
}

std::pair<KDNode*, double> findNearestNeighbor(KDNode* root, const Point& target, int depth) {
    if (root == nullptr) {
        return std::make_pair(nullptr, std::numeric_limits<double>::max());
    }

    int axis = depth % 2;
    KDNode* nextBranch = (axis == 0 && target.x < root->point.x) || (axis == 1 && target.y < root->point.y)
                             ? root->left
                             : root->right;

    auto [bestNode, bestDistance] = findNearestNeighbor(nextBranch, target, depth + 1);

    double currentDistance = distance(target, root->point);
    if (currentDistance < bestDistance) {
        bestDistance = currentDistance;
        bestNode = root;
    }

    double axisDistance = (axis == 0) ? std::abs(target.x - root->point.x) : std::abs(target.y - root->point.y);

    if (axisDistance < bestDistance) {
        KDNode* otherBranch = (nextBranch == root->left) ? root->right : root->left;
        auto [otherBest, otherDistance] = findNearestNeighbor(otherBranch, target, depth + 1);

        if (otherBest != nullptr && otherDistance < bestDistance) {
            bestDistance = otherDistance;
            bestNode = otherBest;
        }
    }

    return std::make_pair(bestNode, bestDistance);
}

std::vector<Point> kMeansKDTree(const std::vector<Point>& data, int k, int maxIterations) {
    // Construir KD-Tree
    KDNode* root = buildKDTree(data, 0);

    // Inicialización de centroides aleatorios sin duplicados
    std::vector<Point> centroids;
    std::vector<int> centroidIndices(data.size());
    std::iota(centroidIndices.begin(), centroidIndices.end(), 0);
    std::random_shuffle(centroidIndices.begin(), centroidIndices.end());

    for (int i = 0; i < k; ++i) {
        centroids.push_back(data[centroidIndices[i]]);
    }

    for (int iteration = 0; iteration < maxIterations; ++iteration) {
        // Asignación de puntos a centroides más cercanos usando el KD-Tree
        for (int i = 0; i < data.size(); ++i) {
            auto [nearestNode, _] = findNearestNeighbor(root, data[i], 0);
            int cluster = std::distance(centroids.begin(), std::find(centroids.begin(), centroids.end(), nearestNode->point));
            centroidIndices[i] = cluster;
        }

        // Actualización de centroides
        std::vector<int> counts(k, 0);
        std::vector<Point> newCentroids(k, Point());
        for (int i = 0; i < data.size(); ++i) {
            int cluster = centroidIndices[i];
            newCentroids[cluster].x += data[i].x;
            newCentroids[cluster].y += data[i].y;
            counts[cluster]++;
        }

        for (int i = 0; i < k; ++i) {
            if (counts[i] > 0) {
                centroids[i].x = newCentroids[i].x / counts[i];
                centroids[i].y = newCentroids[i].y / counts[i];
            }
        }
    }

    // Liberar memoria del KD-Tree
    delete root;

    return centroids;
}

int main() {
    // Lectura del archivo CSV
    std::ifstream file("data2k.csv");
    if (!file.is_open()) {
        std::cerr << "Error al abrir el archivo." << std::endl;
        return 1;
    }

    std::vector<Point> data;
    std::string line;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        double x, y;
        char comma;
        if (iss >> x >> comma >> y) {
            data.emplace_back(x, y);
        } else {
            //std::cerr << "Error al leer la línea: " << line << std::endl;
        }
    }
    file.close();

    // Parámetros de kMeansKDTree
    int k =3;
    int maxIterations = 5;

    // Aplicar kMeansKDTree
    std::vector<Point> centroids = kMeansKDTree(data, k, maxIterations);

    // Imprimir los centroides
    std::cout << "Centroides:\n";
    for (const auto& centroid : centroids) {
        std::cout << "Centroide: (" << centroid.x << ", " << centroid.y << ")\n";
    }

    return 0;
}

File: kmeans_kdtree.py
import csv
import random
import matplotlib.pyplot as plt
import numpy as np

class KDNode:
    def __init__(self, point, left=None, right=None, axis=None):
        self.point = point
        self.left = left
        self.right = right
        self.axis = axis

def build_kdtree(points, depth=0):
    if not points:
        return None
    k = len(points[0])
    axis = depth % k
    points.sort(key=lambda x: x[axis])
    median = len(points) // 2
    return KDNode(
        point=points[median],
        left=build_kdtree(points[:median], depth + 1),
        right=build_kdtree(points[median + 1:], depth + 1),
        axis=axis
    )

def euclidean_distance(point1, point2):
    return np.linalg.norm(np.array(point1) - np.array(point2))

def closest_point(root, target, depth=0, best=None, tol=1e-5):
    if root is None:
        return best

    k = len(target)
    axis = depth % k
    next_best = None
    next_branch = None

    if best is None or euclidean_distance(target, best) > euclidean_distance(target, root.point):
        next_best = root.point
    else:
        next_best = best

    if target[axis] < root.point[axis]:
        next_branch = root.left
    else:
        next_branch = root.right

    if np.allclose(target, next_best, atol=tol):
        return next_best
    else:
        return closest_point(next_branch, target, depth + 1, next_best, tol)

def kmeans_kdtree(points, k, iterations):
    kdtree = build_kdtree(points)
    centroids = random.sample(points, k)

    for _ in range(iterations):
        clusters = {i: [] for i in range(k)}
        for point in points:
            closest_centroid = closest_point(kdtree, point, tol=1e-5)

            # Encontrar el índice del centroide más cercano
            cluster_index = None
            for i, centroid in enumerate(centroids):
                if np.allclose(centroid, closest_centroid, atol=1e-5):
                    cluster_index = i
                    break

            # Asignar el punto al cluster correspondiente
            if cluster_index is not None:
                clusters[cluster_index].append(point)

        # Actualizar los centroides
        for i in range(k):
            if clusters[i]:
                centroids[i] = np.mean(clusters[i], axis=0)

        # Imprimir puntos finales de todos los centroides
        print(f"Iteration {_ + 1}: Centroides finales")
        for i, centroid in enumerate(centroids):
            print(f"Centroide {i + 1}: {centroid}")
        print()

        # Imprimir todos los puntos
        print(f"Iteration {_ + 1}: All Points")
        print(np.array(points))
        print()

        # Graficar todos los puntos y centroides
        plot_clusters_all_points(points, centroids, clusters, _ + 1)

# Función para graficar todos los puntos y centroides
def plot_clusters_all_points(points, centroids, clusters, iteration):
    colors = ['r', 'g', 'b', 'c', 'm', 'y', 'k']
    plt.figure()

    # Graficar todos los puntos
    all_points = np.array(points)
    plt.scatter(all_points[:, 0], all_points[:, 1], c='gray', alpha=0.5, label='All Points')

    # Graficar centroides
    centroids = np.array(centroids)
    plt.scatter(centroids[:, 0], centroids[:, 1], c='black', marker='X', label='Centroids')

    # Graficar clusters
    for i, cluster_points in clusters.items():
        cluster_points = np.array(cluster_points)
        color_index = i % len(colors)  # Ajustar el índice para evitar desbordamiento
        plt.scatter(cluster_points[:, 0], cluster_points[:, 1], c=colors[color_index], label=f'Cluster {i + 1}')

    plt.title(f'K-means Clustering - Iteration {iteration}')
    plt.legend()
    plt.show()

# Leer datos desde el archivo CSV
data = []
with open('data2k.csv', 'r') as file:
    reader = csv.reader(file)
    for row in reader:
        data.append([float(row[0]), float(row[1])])

# Parámetros del algoritmo
k_value = 18
num_iterations = 10

# Ejecutar K-means con KD-tree
kmeans_kdtree(data, k_value, num_iterations)
